### PR TITLE
Circular queue feature

### PR DESCRIPTION
**Pull request by navin-varu**
Please include a summary of the change. Please also include relevant motivation and context (if applicable).

Hi,

I have implemented Circular Queue. 

It is a linear data structure in which the operations are performed based on FIFO principle and last position is connected back to the first position to make a circle. It is also called ‘Ring Buffer’. 

In a normal Queue, we can insert elements until queue becomes full. But once queue becomes full, we can not insert the next element even if there is a space in front of queue. However in circular queue, we can handle this efficiently and use  unused memory locations.

Usages:

1. Memory Management
2. Traffic system
3. CPU Scheduling

- [x] I have performed a self-review of my code
- [x] My code follows the style guidelines of this project
- [x] I have added comments to hard-to-understand areas of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the comments
- [x] I have made corresponding changes to the README.md

### CODE REVIEW

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 18
           

Complexity increasing per file
==============================
- DataStructures.Tests/ArrayBasedCircularQueueTests.cs  2
- DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs  3
- DataStructures/ArrayBasedCircularQueue/CircularQueue.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193&bid=20492863)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
+    {
+        private readonly T[] queue;
+        private bool disposedValue = false;
+        private int front = 0;
+        private int rear = 0;
+        private int position;
+        private Func<int, int> incrementer;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:19]: Codacy found an issue: [Make 'incrementer' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            this.position = this.front - 1;
+        }
+
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(true);
+
+            // TODO: uncomment the following line if the finalizer is overridden above.
+            // GC.SuppressFinalize(this);
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:74]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
+                }
+
+                // TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:94]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            this.position = this.front - 1;
+        }
+
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(true);
+
+            // TODO: uncomment the following line if the finalizer is overridden above.
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:73]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        // }
+
+        // This code added to correctly implement the disposable pattern.
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:91]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/CircularQueue.cs [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    /// <para>Memory Management.</para>
+    /// <para>Traffic system.</para>
+    /// <para>CPU Scheduling.</para>
+    /// </remarks>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class CircularQueue<T> : IQueueOperations<T>
+    {
+        private readonly int size = 10;
+        private int front = -1;
+        private int rear = -1;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/CircularQueue.cs:24]: Codacy found an issue: [Remove the member initializer, all constructors set an initial value for the member.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(true);
+
+            // TODO: uncomment the following line if the finalizer is overridden above.
+            // GC.SuppressFinalize(this);
+        }
+
+        // TODO: override a finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
+        // ~QueueEnumerator()
+        // {
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:79]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/CircularQueue.cs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    /// <para>Usages:</para>
+    /// <para>Memory Management.</para>
+    /// <para>Traffic system.</para>
+    /// <para>CPU Scheduling.</para>
+    /// </remarks>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class CircularQueue<T> : IQueueOperations<T>
+    {
+        private readonly int size = 10;
+        private int front = -1;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/CircularQueue.cs:23]: Codacy found an issue: [Remove the member initializer, all constructors set an initial value for the member.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in Dispose(bool disposing) above.
+            Dispose(true);
+
+            // TODO: uncomment the following line if the finalizer is overridden above.
+            // GC.SuppressFinalize(this);
+        }
+
+        // TODO: override a finalizer only if Dispose(bool disposing) above has code to free unmanaged resources.
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:77]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+namespace DataStructures.ArrayBasedCircularQueue.Enumerators
+{
+    /// <summary>
+    /// Queue Enumerator to enumerate queue.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
+    {
+        private readonly T[] queue;
+        private bool disposedValue = false;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:15]: Codacy found an issue: [Remove this initialization to 'disposedValue', the compiler will do that for you.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures.Tests/ArrayBasedCircularQueueTests.cs [hunk 11]
```diff
@@ -0,0 +1,178 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using DataStructures.ArrayBasedCircularQueue;
+using NUnit.Framework;
+
+namespace DataStructures.Tests
+{
+    public class ArrayBasedCircularQueueTests
```
Comment (siriak) [inline DataStructures.Tests/ArrayBasedCircularQueueTests.cs:9]: Codacy found an issue: [Add a 'protected' constructor or the 'static' keyword to the class declaration.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        private int front = 0;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:16]: Codacy found an issue: [Remove this initialization to 'front', the compiler will do that for you.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
+                    // TODO: dispose managed state (managed objects).
+                }
+
+                // TODO: free unmanaged resources (unmanaged objects) and override a finalizer below.
+                // TODO: set large fields to null.
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:95]: Codacy found an issue: [Complete the task associated to this 'TODO' comment.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    /// <summary>
+    /// Queue Enumerator to enumerate queue.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
+    {
+        private readonly T[] queue;
+        private bool disposedValue = false;
+        private int front = 0;
+        private int rear = 0;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:17]: Codacy found an issue: [Remove this initialization to 'rear', the compiler will do that for you.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        private int position;
+        private Func<int, int> incrementer;
+        private Func<int, int, bool> condition;
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:20]: Codacy found an issue: [Make 'condition' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs (hunk 12 above)
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:16]: Codacy found an issue: [Make 'front' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace DataStructures.ArrayBasedCircularQueue.Enumerators
+{
+    /// <summary>
+    /// Queue Enumerator to enumerate queue.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:12]: Codacy found an issue: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs (hunk 14 above)
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:17]: Codacy found an issue: [Make 'rear' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- DataStructures.Tests/ArrayBasedCircularQueueTests.cs  2
- DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs  3
- DataStructures/ArrayBasedCircularQueue/CircularQueue.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193&bid=20492863)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (!disposedValue)
+            {
+                if (disposing)
+                {
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:78]: Codacy found an issue: [Either remove or fill this block of code.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace DataStructures.ArrayBasedCircularQueue.Enumerators
+{
+    /// <summary>
+    /// Queue Enumerator to enumerate queue.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:12]: Codacy found an issue: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- DataStructures.Tests/ArrayBasedCircularQueueTests.cs  2
- DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs  1
- DataStructures/ArrayBasedCircularQueue/CircularQueue.cs  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193&bid=20492863)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+        /// <summary>
+        /// Resets enumerator position.
+        /// </summary>
+        public void Reset()
+        {
+            this.position = this.front - 1;
+        }
+
+        public void Dispose()
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:67]: Codacy found an issue: [Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace DataStructures.ArrayBasedCircularQueue.Enumerators
+{
+    /// <summary>
+    /// Queue Enumerator to enumerate queue.
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class QueueEnumerator<T> : IEnumerator<T>
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:12]: Codacy found an issue: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs  1
- DataStructures/ArrayBasedCircularQueue/CircularQueue.cs  3
- DataStructures.Tests/ArrayBasedCircularQueueTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193&bid=20492863)

File: DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
```
Comment (siriak) [inline DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs:12]: Codacy found an issue: [Fix this implementation of 'IDisposable' to conform to the dispose pattern.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures/ArrayBasedCircularQueue/Enumerators/QueueEnumerator.cs  1
- DataStructures/ArrayBasedCircularQueue/CircularQueue.cs  3
- DataStructures.Tests/ArrayBasedCircularQueueTests.cs  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6301193&bid=20492863)

**Review by siriak — COMMENTED** [review]
Hi! There is an ArrayBasedQueue class, which is basically the same data structure as ArrayBasedCircularQueue. Could you implement your improvements on top of it? I think that having an enumerator and a common interface for all queue types would be good. And also, having two implementations of the same data structure without significant differences is not good.

**Comment by navin-varu** [comment]
Hi, 

I agree with your concern.

Do you want me to implement enumerator for Arraybasedqueue as well and abstract interface for that queue as well?

Is it ok if I will do some changes in ArrayBasedQueue?


Proposed changes:

ArrayBasedQueue will implement IQueryOperations interface as well as IEnumerable interface.

Please confirm so that I can go ahead.

**Comment by navin-varu** [comment]
Hi,

I have rechecked my implementation, I feel even though interfaces are same for both the queues, but implementation details are different. Because we can add element in circular queue from front and rear if it is not full.

Can you please suggest in this regards?

**Comment by siriak** [comment]
> Do you want me to implement enumerator for Arraybasedqueue as well and abstract interface for that queue as well?

Yes

> Is it ok if I will do some changes in ArrayBasedQueue?

Yes, if they improve something

> ArrayBasedQueue will implement IQueryOperations interface as well as IEnumerable interface.

Yes, that sounds about correct

> I have rechecked my implementation, I feel even though interfaces are same for both the queues, but implementation details are different. Because we can add element in circular queue from front and rear if it is not full.

I don't really understand what you are saying. They both are based on array T[] and work with it as with a circular array (write in the beginning when the end is reached and all that stuff)

**Comment by siriak** [comment]
I'm closing this one, feel free to reopen it or create a new one as soon as the discussed findings are fixed

**Pull request closed without merging**